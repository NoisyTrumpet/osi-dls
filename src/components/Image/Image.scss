.lazy-image-container {
  opacity: 0;
  transition: opacity $timing-2;
  &.is-visible {
    opacity: 1;
  }
}
