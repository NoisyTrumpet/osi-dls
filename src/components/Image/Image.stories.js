import React from 'react';
import { storiesOf } from '@storybook/react';
import { withSmartKnobs } from 'storybook-addon-smart-knobs';
import { Image } from './';

const parameters = {
  component: Image,
  componentSubtitle: 'The <Image/> tag defines a <img> tag with optional sources. This component does not render children.',
  jest: ['Image'],
};

storiesOf('components/Image', module)
  .addParameters(parameters)
  .addDecorator(withSmartKnobs({ ignoreProps: ['onClick', 'lazyLoad'] }))
  .add('Default', (() => <Image src="https://www.noisytrumpet.com/wp-content/uploads/2021/01/osi.png" alt="OSI-LIFE" />));

storiesOf('components/Image', module)
  .addParameters(parameters)
  .add('Mulitple Sources (Responsive)', () => (
    <Image
      src="https://www.w3schools.com/tags/img_pink_flowers.jpg"
      alt="Flowers"
    >
      <Image.Source media="(max-width:700px)" srcSet="https://www.w3schools.com/tags/img_orange_flowers.jpg" />
      <Image.Source media="(max-width:900px)" srcSet="https://www.w3schools.com/tags/img_white_flower.jpg" />
    </Image>
  ));

storiesOf('components/Image', module)
  .addParameters(parameters)
  .add('Lazy Loading (on by default)', (() => (
    <>
      <div style={{ height: '600px' }}>
        <Image src="https://www.noisytrumpet.com/wp-content/uploads/2021/01/osi.png" alt="Image 1" />
      </div>
      <div style={{ height: '600px' }}>
        <Image src="https://www.noisytrumpet.com/wp-content/uploads/2021/01/osi.png" alt="Image 2" />
      </div>
      <div style={{ height: '600px' }}>
        <Image src="https://www.noisytrumpet.com/wp-content/uploads/2021/01/osi.png" alt="Image 3" />
      </div>
      <div style={{ height: '600px' }}>
        <Image src="https://www.noisytrumpet.com/wp-content/uploads/2021/01/osi.png" alt="Image 4" />
      </div>
    </>
  )));
