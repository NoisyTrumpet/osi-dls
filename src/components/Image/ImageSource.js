import React from 'react';
import PropTypes from 'prop-types';

const ImageSource = ({ srcSet, dataSrcSet, media }) => (
  <source srcSet={srcSet} data-srcset={dataSrcSet} media={media} />
);

ImageSource.defaultProps = {};

ImageSource.propTypes = {
  media: PropTypes.string,
  srcSet: PropTypes.string,
  dataSrcSet: PropTypes.string,
};

export default ImageSource;
