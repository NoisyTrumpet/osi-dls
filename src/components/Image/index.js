export { default as Image } from './Image'; // eslint-disable-line import/prefer-default-export
