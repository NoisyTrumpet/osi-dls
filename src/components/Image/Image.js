import React from 'react';
import PropTypes from 'prop-types';
import LazyLoad from 'react-lazy-load'; // https://www.npmjs.com/package/react-lazy-load
import classNames from 'classnames';
import ImageSource from './ImageSource';
import './Image.scss';

const Image = ({
  src,
  dataSrc,
  alt,
  className,
  addClass,
  children,
  pictureClass,
  ariaHidden,
  lazyLoad,
  lazyOffsetVertical,
  lazyOffsetHorizontal,
}) => {
  const classes = classNames(
    className || 'image',
    addClass,
  );

  const imageComponent = (
    <picture className={pictureClass}>
      {children}
      <img data-src={dataSrc} src={src} alt={alt} className={classes} aria-hidden={ariaHidden} />
    </picture>
  );

  return (
    lazyLoad ? <LazyLoad className="lazy-image-container" debounce={false} offsetHorizontal={lazyOffsetHorizontal} offsetVertical={lazyOffsetVertical}>{imageComponent}</LazyLoad> : <>{imageComponent}</>
  );
};

Image.Source = ImageSource;

Image.defaultProps = {
  lazyLoad: true,
  lazyOffsetVertical: 128,
  lazyOffsetHorizontal: 128,
};

Image.propTypes = {
  src: PropTypes.string.isRequired,
  alt: PropTypes.string.isRequired,
  dataSrc: PropTypes.string,
  className: PropTypes.string,
  addClass: PropTypes.string,
  pictureClass: PropTypes.string,
  ariaHidden: PropTypes.string,
  /** Flag to determine if image should be lazily loaded (Gets wrapped with https://www.npmjs.com/package/react-lazy-load) */
  lazyLoad: PropTypes.bool,
  /** https://www.npmjs.com/package/react-lazy-load#offsetvertical */
  lazyOffsetVertical: PropTypes.number,
  /** https://www.npmjs.com/package/react-lazy-load#offsethorizontal */
  lazyOffsetHorizontal: PropTypes.number,
  children: (props, propName, componentName) => {
    let error;
    const prop = props[propName];
    const requiredChildType = 'ImageSource';
    React.Children.forEach(prop, (child) => {
      // type.name seems to work for both Class and Functional components
      if (child.type.name !== requiredChildType) {
        error = new Error(`\`${componentName}\` only accepts children of type \`${requiredChildType}\`. Child of type ${child.type.name} was found.`);
      }
    });
    return error;
  },
};

export default Image;
