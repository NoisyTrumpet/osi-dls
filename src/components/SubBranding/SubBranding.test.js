import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { SubBranding } from './';
import { Image } from '../Image';

describe('<SubBranding />', () => {
  test('should render ', () => {
    const wrapper = shallow(<SubBranding src="https://webfileswest.uswest.os.ctl.io/corporate.hallmark.com/wp-content/uploads/2018/06/Hallmark-Greetings-Logo-black-160x160.jpg" alt="Hallmark" />);
    expect(wrapper.find(Image).length).toBe(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
