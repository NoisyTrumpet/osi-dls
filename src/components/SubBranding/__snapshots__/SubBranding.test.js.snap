// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubBranding /> should render  1`] = `
<Image
  alt="Hallmark"
  className="sub-branding"
  lazyLoad={true}
  lazyOffsetHorizontal={128}
  lazyOffsetVertical={128}
  src="https://webfileswest.uswest.os.ctl.io/corporate.hallmark.com/wp-content/uploads/2018/06/Hallmark-Greetings-Logo-black-160x160.jpg"
/>
`;
